File: articles/reaction-networks-chemical-systems-modeling/julia/reaction_network_kernel.jl
# Reaction-network numerical kernel in Julia.

function simulate_network(k1, k2, k3, k4, dt, total_time)
    A = 1.0
    B = 0.0
    C = 0.0
    D = 0.0
    E = 0.0

    for t in 0:dt:total_time
        r1 = k1 * A
        r2 = k2 * B
        r3 = k3 * A
        r4 = k4 * B

        A = max(A + (-r1 - r3) * dt, 0.0)
        B = max(B + (r1 - r2 - r4) * dt, 0.0)
        C = max(C + r2 * dt, 0.0)
        D = max(D + r3 * dt, 0.0)
        E = max(E + r4 * dt, 0.0)
    end

    return A, B, C, D, E
end

A, B, C, D, E = simulate_network(0.20, 0.08, 0.05, 0.03, 0.25, 50.0)

println("A_final=", round(A, digits=6))
println("B_final=", round(B, digits=6))
println("C_final=", round(C, digits=6))
println("D_final=", round(D, digits=6))
println("E_final=", round(E, digits=6))

File: articles/reaction-networks-chemical-systems-modeling/fortran/reaction_network_kernel.f90
program reaction_network_kernel
  implicit none

  real(8) :: A, B, C, D, E
  real(8) :: k1, k2, k3, k4, dt, t, total_time
  real(8) :: r1, r2, r3, r4

  A = 1.0d0
  B = 0.0d0
  C = 0.0d0
  D = 0.0d0
  E = 0.0d0

  k1 = 0.20d0
  k2 = 0.08d0
  k3 = 0.05d0
  k4 = 0.03d0
  dt = 0.25d0
  total_time = 50.0d0

  t = 0.0d0
  do while (t <= total_time)
    r1 = k1 * A
    r2 = k2 * B
    r3 = k3 * A
    r4 = k4 * B

    A = max(A + (-r1 - r3) * dt, 0.0d0)
    B = max(B + (r1 - r2 - r4) * dt, 0.0d0)
    C = max(C + r2 * dt, 0.0d0)
    D = max(D + r3 * dt, 0.0d0)
    E = max(E + r4 * dt, 0.0d0)

    t = t + dt
  end do

  print *, "A_final=", A
  print *, "B_final=", B
  print *, "C_final=", C
  print *, "D_final=", D
  print *, "E_final=", E

end program reaction_network_kernel
